### ISSUE TITLE
msan: use of uninitialized value in lowerUTF8()

### ISSUE THREAD

**Issue by rschu1ze** [ClickHouse/ClickHouse#51314]
https://s3.amazonaws.com/clickhouse-test-reports/51288/1242a1fd95dd7aaaa80c8c4ba49a290db5cd1a02/fuzzer_astfuzzermsan/report.html

`SELECT lower(materialize('')) = materialize('aaaaАБВГaaaaaaaa'), lowerUTF8(materialize('aaaaАБВГAAAAaaAAaaaaАБВГAAAAaaAA')) = materialize(NULL), 255 FROM test__fuzz_0`

```
2023.06.23 00:46:35.630741 [ 515 ] {} <Fatal> BaseDaemon: ########################################
2023.06.23 00:46:35.631301 [ 515 ] {} <Fatal> BaseDaemon: (version 23.6.1.1, build id: E513C8D69E0A119FD25EFF30C39852E2669A1033, git hash: 841c8188f6d089e1d86e8f2aa28cf88a2b9ab35a) (from thread 505) (query_id: ecb3c2d4-ad95-4012-af52-608fceb01aeb) (query: SELECT lower(materialize('')) = materialize('aaaaАБВГaaaaaaaa'), lowerUTF8(materialize('aaaaАБВГAAAAaaAAaaaaАБВГAAAAaaAA')) = materialize(NULL), 255 FROM test__fuzz_0) Received signal sanitizer trap (-3)
2023.06.23 00:46:35.631536 [ 515 ] {} <Fatal> BaseDaemon: Sanitizer trap.
2023.06.23 00:46:35.631754 [ 515 ] {} <Fatal> BaseDaemon: Stack trace: 0x000055e0da5524bb 0x000055e0dad2527d 0x000055e0dad1e3af 0x000055e0c1fec576 0x000055e0c2001063 0x000055e0cfe8be56 0x000055e0cfe8a670 0x000055e0c207a838 0x000055e0ee87075d 0x000055e0ee872665 0x000055e0ee8783d8 0x000055e0f0a8c081 0x000055e0f7e8e13b 0x000055e0e4574341 0x000055e0f756f4de 0x000055e0f75d3381 0x000055e0f75a683d 0x000055e0f75a4fde 0x000055e0f75a4a49 0x000055e0f75def77 0x000055e0da7d345f 0x000055e0da7e0b2b 0x00007f2c77d74609 0x00007f2c77c99133
2023.06.23 00:46:35.738367 [ 515 ] {} <Fatal> BaseDaemon: 0. ./build_docker/./src/Common/StackTrace.cpp:291: StackTrace::tryCapture() @ 0x0000000027b064bb in /workspace/clickhouse
2023.06.23 00:46:36.005895 [ 515 ] {} <Fatal> BaseDaemon: 1. ./build_docker/./src/Common/StackTrace.h:49: StackTrace::StackTrace() @ 0x00000000282d927d in /workspace/clickhouse
2023.06.23 00:46:36.271745 [ 515 ] {} <Fatal> BaseDaemon: 2. ./build_docker/./src/Daemon/BaseDaemon.cpp:486: sanitizerDeathCallback() @ 0x00000000282d23af in /workspace/clickhouse
2023.06.23 00:46:41.188779 [ 515 ] {} <Fatal> BaseDaemon: 3. __sanitizer::Die() @ 0x000000000f5a0576 in /workspace/clickhouse
2023.06.23 00:46:42.921195 [ 516 ] {} <Fatal> BaseDaemon: ########################################
2023.06.23 00:46:42.921325 [ 516 ] {} <Fatal> BaseDaemon: (version 23.6.1.1, build id: E513C8D69E0A119FD25EFF30C39852E2669A1033, git hash: 841c8188f6d089e1d86e8f2aa28cf88a2b9ab35a) (from thread 505) (query_id: ecb3c2d4-ad95-4012-af52-608fceb01aeb) (query: SELECT lower(materialize('')) = materialize('aaaaАБВГaaaaaaaa'), lowerUTF8(materialize('aaaaАБВГAAAAaaAAaaaaАБВГAAAAaaAA')) = materialize(NULL), 255 FROM test__fuzz_0) Received signal sanitizer trap (-3)
2023.06.23 00:46:42.921404 [ 516 ] {} <Fatal> BaseDaemon: Sanitizer trap.
2023.06.23 00:46:42.921478 [ 516 ] {} <Fatal> BaseDaemon: Stack trace: 0x000055e0da5524bb 0x000055e0dad2527d 0x000055e0dad1e3af 0x000055e0c1fec576 0x000055e0c2011f5f 0x000055e0da67db46 0x000055e0da930db2 0x000055e0dad1e419 0x000055e0c1fec576 0x000055e0c2001063 0x000055e0cfe8be56 0x000055e0cfe8a670 0x000055e0c207a838 0x000055e0ee87075d 0x000055e0ee872665 0x000055e0ee8783d8 0x000055e0f0a8c081 0x000055e0f7e8e13b 0x000055e0e4574341 0x000055e0f756f4de 0x000055e0f75d3381 0x000055e0f75a683d 0x000055e0f75a4fde 0x000055e0f75a4a49 0x000055e0f75def77 0x000055e0da7d345f 0x000055e0da7e0b2b 0x00007f2c77d74609 0x00007f2c77c99133
2023.06.23 00:46:43.029217 [ 516 ] {} <Fatal> BaseDaemon: 0. ./build_docker/./src/Common/StackTrace.cpp:291: StackTrace::tryCapture() @ 0x0000000027b064bb in /workspace/clickhouse
2023.06.23 00:46:43.295389 [ 516 ] {} <Fatal> BaseDaemon: 1. ./build_docker/./src/Common/StackTrace.h:49: StackTrace::StackTrace() @ 0x00000000282d927d in /workspace/clickhouse
2023.06.23 00:46:43.560882 [ 516 ] {} <Fatal> BaseDaemon: 2. ./build_docker/./src/Daemon/BaseDaemon.cpp:486: sanitizerDeathCallback() @ 0x00000000282d23af in /workspace/clickhouse
2023.06.23 00:46:45.979437 [ 515 ] {} <Fatal> BaseDaemon: 4. ? @ 0x000000000f5b5063 in /workspace/clickhouse
2023.06.23 00:46:48.374624 [ 516 ] {} <Fatal> BaseDaemon: 3. __sanitizer::Die() @ 0x000000000f5a0576 in /workspace/clickhouse
2023.06.23 00:46:50.774525 [ 515 ] {} <Fatal> BaseDaemon: 5. DB::LowerUpperUTF8Impl<(char)65, (char)90, &Poco::Unicode::toLower(int), &void DB::UTF8CyrillicToCase<true>(char8_t const*&, char8_t*&)>::array(char8_t const*, char8_t const*, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 63ul, 64ul> const&, char8_t*) @ 0x000000001d43fe56 in /workspace/clickhouse
2023.06.23 00:46:53.201696 [ 516 ] {} <Fatal> BaseDaemon: 4. ? @ 0x000000000f5c5f5f in /workspace/clickhouse
2023.06.23 00:46:53.207755 [ 516 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/IO/WriteBufferFromFileDescriptorDiscardOnFailure.cpp:16: DB::WriteBufferFromFileDescriptorDiscardOnFailure::nextImpl() @ 0x0000000027c31b46 in /workspace/clickhouse
2023.06.23 00:46:53.641523 [ 516 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/IO/WriteBuffer.h:61: DB::WriteBuffer::next() @ 0x0000000027ee4db2 in /workspace/clickhouse
2023.06.23 00:46:53.907062 [ 516 ] {} <Fatal> BaseDaemon: 7. ./build_docker/./src/Daemon/BaseDaemon.cpp:495: sanitizerDeathCallback() @ 0x00000000282d2419 in /workspace/clickhouse
2023.06.23 00:46:55.573115 [ 515 ] {} <Fatal> BaseDaemon: 6. DB::FunctionStringToString<DB::LowerUpperUTF8Impl<(char)65, (char)90, &Poco::Unicode::toLower(int), &void DB::UTF8CyrillicToCase<true>(char8_t const*&, char8_t*&)>, DB::(anonymous namespace)::NameLowerUTF8, false>::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x000000001d43e670 in /workspace/clickhouse
2023.06.23 00:46:58.722836 [ 516 ] {} <Fatal> BaseDaemon: 8. __sanitizer::Die() @ 0x000000000f5a0576 in /workspace/clickhouse
2023.06.23 00:47:00.367207 [ 515 ] {} <Fatal> BaseDaemon: 7. DB::FunctionToExecutableFunctionAdaptor::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x000000000f62e838 in /workspace/clickhouse
2023.06.23 00:47:00.484865 [ 515 ] {} <Fatal> BaseDaemon: 8. ./build_docker/./src/Functions/IFunction.cpp:0: DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x000000003be2475d in /workspace/clickhouse
```

**Comment by rschu1ze** [comment]
Minimal repro: `SELECT lowerUTF8(materialize('aaaaАБВГAAAAaaAAaaaaАБВГAAAAaaAA')) from tab`

**Comment by alexey-milovidov** [comment]
While they could read, process, and write uninitialized memory, they can maximally write to unused column's padding.
Changing from "major" to "fuzz".

**Comment by rschu1ze** [comment]
Fixed by #51371.
